lib/puppet/util/feature.rb

```diff
@@ -3,39 +3,49 @@
 
-  # Create a new feature test.  You have to pass the feature name,
-  # and it must be unique.  You can either provide a block that
-  # will get executed immediately to determine if the feature
-  # is present, or you can pass an option to determine it.
-  # Currently, the only supported option is 'libs' (must be
-  # passed as a symbol), which will make sure that each lib loads
-  # successfully.
-  def add(name, options = {})
+  # Create a new feature test. You have to pass the feature name, and it must be
+  # unique. You can pass a block to determine if the feature is present:
+  #
+  #     Puppet.features.add(:myfeature) do
+  #       # return true or false if feature is available
+  #     end
+  #
+  # The block should return true if the feature is available, false if it is
+  # not, or nil if the state is unknown. True and false values will be cached. A
+  # nil value will not be cached, and should be used if the feature may become
+  # true in the future.
+  #
+  # Feature are often used to detect if a ruby library is installed. To support
+  # that common case, you can pass one or more ruby libraries, and the feature
+  # will be true if all of the libraries load successfully:
+  #
+  #     Puppet.features.add(:myfeature, libs: 'mylib')
+  #     Puppet.features.add(:myfeature, libs: ['mylib', 'myotherlib'])
+  #
+  # If the ruby library is not installed, then the failure is not cached, as
+  # it's assumed puppet may install the gem during catalog application.
+  #
+  # Puppet evaluates the feature test when the `Puppet.features.myfeature?`
+  # method is called. If the feature test was defined using a block and the
+  # block returns nil, then the feature test will be re-evaluated the next time
+  # `Puppet.features.myfeature?` is called.
+  #
+  # @param [Symbol] name The unique feature name
+  # @param [Hash<Symbol,Array<String>>] options The libraries to load
+  def add(name, options = {}, &block)
     method = name.to_s + "?"
     @results.delete(name)
 
-    if block_given?
-      begin
-        result = yield
-      rescue StandardError,ScriptError => detail
-        warn _("Failed to load feature test for %{name}: %{detail}") % { name: name, detail: detail }
-        result = false
-      end
-      @results[name] = result
-    end
-
     meta_def(method) do
       # we return a cached result if:
-      #  * if a block is given (and we just evaluated it above)
       #  * if we already have a positive result
       #  * if we've tested this feature before and it failed, but we're
       #    configured to always cache
-      if block_given?     ||
-          @results[name]  ||
-          (@results.has_key?(name) && (!Puppet[:always_retry_plugins]))
+      if !@results[name].nil?  ||
+         (@results.has_key?(name) && (!Puppet[:always_retry_plugins]))
         @results[name]
       else
-        @results[name] = test(name, options)
+        @results[name] = test(name, options, &block)
         @results[name]
       end
```

Comment:
`Puppet.features.<name>?` should always return true or false, not truthy or nil.

Reply:
```
irb(main):002:0> Puppet.features.add(:foo) do nil end
=> :foo?
irb(main):003:0> Puppet.features.foo?
TESTING foo
=> nil
irb(main):004:0> Puppet.features.add(:foo) do Integer end
=> :foo?
irb(main):005:0> Puppet.features.foo?
TESTING foo
=> Integer
```